File: go/day13/day13.go
package day13

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/tadejsv/aoc-2024/utils"
)

type Machine struct {
	aVals [2]int64
	bVals [2]int64
	pVals [2]int64
}

func gcd_euclid(a, b int64) int64 {
	if b == 0 {
		return a
	}
	return gcd_euclid(b, a%b)
}

func extended_euclidean(a, b int64) (x, y, gcd int64) {
	if b == 0 {
		return 1, 0, a
	}
	xOld, yOld, gcd := extended_euclidean(b, a%b)
	x = yOld
	y = xOld - yOld*(a/b)
	return x, y, gcd
}

func solveExtra(m *Machine) int64 {
	const costA = 3
	const costB = 1

	// Check for co-linear equations - assume they all have no solutions (true for given inputs)
	det := m.bVals[1]*m.aVals[0] - m.bVals[0]*m.aVals[1]
	var B int64
	if det == 0 {
		// Test for inconsistency
		if m.aVals[0]/m.aVals[1] != m.pVals[0]/m.pVals[1] {
			return 0
		}

		// Test for lack of integer solutions
		if m.pVals[0]%gcd_euclid(m.aVals[0], m.bVals[0]) != 0 {
			return 0
		}

		aPrime := m.aVals[0] / gcd_euclid(m.aVals[0], m.bVals[0])
		bInv, _, _ := extended_euclidean(m.bVals[0], aPrime)
		bMod := (m.pVals[0] * bInv) % aPrime
		if bMod < 0 {
			bMod += aPrime
		}

		if float64(costA)/float64(costB) > float64(m.aVals[0])/float64(m.bVals[0]) {
			bMax := m.pVals[0] / m.bVals[0]
			B = aPrime*(bMax/aPrime) + bMod
			if B > bMax {
				B -= aPrime
			}
		} else {
			B = bMod
		}
	} else {
		cPart := m.pVals[1]*m.aVals[0] - m.pVals[0]*m.aVals[1]
		B = cPart / det
	}
	A := (m.pVals[0] - B*m.bVals[0]) / m.aVals[0]
	if A < 0 || B < 0 {
		return 0
	}

	// Test that we indeed have an integer solution
	pSolution1 := (m.aVals[0] * A) + (m.bVals[0] * B)
	pSolution2 := (m.aVals[1] * A) + (m.bVals[1] * B)

	if pSolution1 == m.pVals[0] && pSolution2 == m.pVals[1] {
		return (3 * A) + B
	}
	return 0
}

func solve(m *Machine) int64 {
	// Check for co-linear equations - assume they all have no solutions (true for given inputs)
	det := m.bVals[1]*m.aVals[0] - m.bVals[0]*m.aVals[1]
	if det == 0 {
		return 0
	}
	cPart := m.pVals[1]*m.aVals[0] - m.pVals[0]*m.aVals[1]
	B := cPart / det
	A := (m.pVals[0] - B*m.bVals[0]) / m.aVals[0]
	if A < 0 || B < 0 {
		return 0
	}

	// Test that we indeed have an integer solution
	pSolution1 := (m.aVals[0] * A) + (m.bVals[0] * B)
	pSolution2 := (m.aVals[1] * A) + (m.bVals[1] * B)

	if pSolution1 == m.pVals[0] && pSolution2 == m.pVals[1] {
		return (3 * A) + B
	}
	return 0
}

func Day13a() {
	lines, err := utils.ReadLines("../inputs/day13.txt")
	if err != nil {
		panic("Could not read lines")
	}

	machines := []Machine{}
	for i := 0; i < len(lines); i += 4 {
		machine := Machine{}
		aParts := strings.Split(strings.Split(lines[i], ": ")[1], ", ")
		bParts := strings.Split(strings.Split(lines[i+1], ": ")[1], ", ")
		pParts := strings.Split(strings.Split(lines[i+2], ": ")[1], ", ")

		for j := range aParts {
			aVal, _ := strconv.Atoi(aParts[j][2:])
			machine.aVals[j] = int64(aVal)
			bVal, _ := strconv.Atoi(bParts[j][2:])
			machine.bVals[j] = int64(bVal)
			pVal, _ := strconv.Atoi(pParts[j][2:])
			machine.pVals[j] = int64(pVal)
		}
		machines = append(machines, machine)
	}

	totalTokens := int64(0)
	for _, machine := range machines {
		totalTokens += solve(&machine)
	}
	fmt.Printf("%v\n", totalTokens)
}

func Day13b() {
	const addPrize = 10000000000000
	lines, err := utils.ReadLines("../inputs/day13.txt")
	if err != nil {
		panic("Could not read lines")
	}

	machines := []Machine{}
	for i := 0; i < len(lines); i += 4 {
		machine := Machine{}
		aParts := strings.Split(strings.Split(lines[i], ": ")[1], ", ")
		bParts := strings.Split(strings.Split(lines[i+1], ": ")[1], ", ")
		pParts := strings.Split(strings.Split(lines[i+2], ": ")[1], ", ")

		for j := range aParts {
			aVal, _ := strconv.Atoi(aParts[j][2:])
			machine.aVals[j] = int64(aVal)
			bVal, _ := strconv.Atoi(bParts[j][2:])
			machine.bVals[j] = int64(bVal)
			pVal, _ := strconv.Atoi(pParts[j][2:])
			machine.pVals[j] = int64(pVal) + addPrize
		}
		machines = append(machines, machine)
	}

	totalTokens := int64(0)
	for _, machine := range machines {
		totalTokens += solve(&machine)
	}
	fmt.Printf("%v\n", totalTokens)
}

func Day13extra() {
	lines, err := utils.ReadLines("../inputs/day13_extra.txt")
	if err != nil {
		panic("Could not read lines")
	}

	machines := []Machine{}
	for i := 0; i < len(lines); i += 4 {
		machine := Machine{}
		aParts := strings.Split(strings.Split(lines[i], ": ")[1], ", ")
		bParts := strings.Split(strings.Split(lines[i+1], ": ")[1], ", ")
		pParts := strings.Split(strings.Split(lines[i+2], ": ")[1], ", ")

		for j := range aParts {
			aVal, _ := strconv.Atoi(aParts[j][2:])
			machine.aVals[j] = int64(aVal)
			bVal, _ := strconv.Atoi(bParts[j][2:])
			machine.bVals[j] = int64(bVal)
			pVal, _ := strconv.Atoi(pParts[j][2:])
			machine.pVals[j] = int64(pVal)
		}
		machines = append(machines, machine)
	}

	totalTokens := int64(0)
	for _, machine := range machines {
		totalTokens += solveExtra(&machine)
	}
	fmt.Printf("%v\n", totalTokens)
}

File: python/day13.py
from dataclasses import dataclass


@dataclass
class Machine:
    button_a: tuple[int, int]
    button_b: tuple[int, int]
    prize: tuple[int, int]


COST_A = 3
COST_B = 1


# Returns x (multitplier of a), y (multipler of b), gcd
def extended_euclidean(a: int, b: int) -> tuple[int, int, int]:
    if b == 0:
        return 1, 0, a

    x_old, y_old, gcd_ret = extended_euclidean(b, a % b)
    x = y_old
    y = x_old - y_old * (a // b)
    return x, y, gcd_ret


def gcd(a: int, b: int) -> int:
    if b == 0:
        return a
    return gcd(b, a % b)


def solve_machine(machine: Machine) -> int:
    det = (
        machine.button_b[1] * machine.button_a[0]
        - machine.button_b[0] * machine.button_a[1]
    )

    if det == 0:
        return 0

    p_part = (
        machine.prize[1] * machine.button_a[0] - machine.prize[0] * machine.button_a[1]
    )
    b = p_part // det

    a = (machine.prize[0] - b * machine.button_b[0]) // machine.button_a[0]

    if a < 0 or b < 0:
        return 0

    p0 = machine.button_a[0] * a + machine.button_b[0] * b
    p1 = machine.button_a[1] * a + machine.button_b[1] * b

    if p0 == machine.prize[0] and p1 == machine.prize[1]:
        return a * COST_A + b * COST_B

    return 0


def solve_machine_extra(machine: Machine) -> int:
    det = (
        machine.button_b[1] * machine.button_a[0]
        - machine.button_b[0] * machine.button_a[1]
    )

    b: int
    if det == 0:
        # Test for inconsistency
        if (
            machine.button_a[0] // machine.button_a[1]
            != machine.prize[0] // machine.prize[1]
        ):
            return 0

        # Test for lack of integer solutions
        if machine.prize[0] % gcd(machine.button_a[0], machine.button_b[0]) != 0:
            return 0

        a_prime = machine.button_a[0] // gcd(machine.button_a[0], machine.button_b[0])
        b_inv, _, _ = extended_euclidean(machine.button_b[0], a_prime)
        b_mod = (machine.prize[0] * b_inv) % a_prime
        if b_mod < 0:
            b_mod += a_prime

        # Using cross-multiplication to avoid casting to float to compute
        # ca / cb > a[0] / b[0]
        if COST_A * machine.button_b[0] > machine.button_a[0] * COST_B:
            b_max = machine.prize[0] // machine.button_b[0]
            b = a_prime * (b_max // a_prime) + b_mod
            if b > b_max:
                b -= a_prime
        else:
            b = b_mod
    else:
        p_part = (
            machine.prize[1] * machine.button_a[0]
            - machine.prize[0] * machine.button_a[1]
        )
        b = p_part // det

    a = (machine.prize[0] - b * machine.button_b[0]) // machine.button_a[0]

    if a < 0 or b < 0:
        return 0

    p0 = machine.button_a[0] * a + machine.button_b[0] * b
    p1 = machine.button_a[1] * a + machine.button_b[1] * b

    if p0 == machine.prize[0] and p1 == machine.prize[1]:
        return a * COST_A + b * COST_B

    return 0


def day13a() -> None:
    machines: list[Machine] = []

    machine = Machine((0, 0), (0, 0), (0, 0))
    with open("../inputs/day13.txt", "r") as file:
        for line in file:
            if len(line.strip()) == 0:
                machines.append(machine)
                machine = Machine((0, 0), (0, 0), (0, 0))
                continue

            parts = line.strip().split(": ")
            if parts[0] == "Prize":
                parts_coords = parts[1].split(", ")
                coords = tuple(int(pc.split("=")[1]) for pc in parts_coords)
                machine.prize = coords
            else:
                parts_coords = parts[1].split(", ")
                coords = tuple(int(pc.split("+")[1]) for pc in parts_coords)

                if parts[0] == "Button A":
                    machine.button_a = coords
                else:
                    machine.button_b = coords

    machines.append(machine)

    total_tokens = 0
    for machine in machines:
        total_tokens += solve_machine(machine)

    print(total_tokens)


def day13b() -> None:
    machines: list[Machine] = []

    machine = Machine((0, 0), (0, 0), (0, 0))
    with open("../inputs/day13.txt", "r") as file:
        for line in file:
            if len(line.strip()) == 0:
                machines.append(machine)
                machine = Machine((0, 0), (0, 0), (0, 0))
                continue

            parts = line.strip().split(": ")
            if parts[0] == "Prize":
                parts_coords = parts[1].split(", ")
                coords = tuple(int(pc.split("=")[1]) for pc in parts_coords)
                add_const = 10000000000000
                machine.prize = (coords[0] + add_const, coords[1] + add_const)
            else:
                parts_coords = parts[1].split(", ")
                coords = tuple(int(pc.split("+")[1]) for pc in parts_coords)

                if parts[0] == "Button A":
                    machine.button_a = coords
                else:
                    machine.button_b = coords

    machines.append(machine)

    total_tokens = 0
    for machine in machines:
        total_tokens += solve_machine(machine)

    print(total_tokens)


def day13extra() -> None:
    machines: list[Machine] = []

    machine = Machine((0, 0), (0, 0), (0, 0))
    with open("../inputs/day13_extra.txt", "r") as file:
        for line in file:
            if len(line.strip()) == 0:
                machines.append(machine)
                machine = Machine((0, 0), (0, 0), (0, 0))
                continue

            parts = line.strip().split(": ")
            if parts[0] == "Prize":
                parts_coords = parts[1].split(", ")
                coords = tuple(int(pc.split("=")[1]) for pc in parts_coords)
                machine.prize = (coords[0], coords[1])
            else:
                parts_coords = parts[1].split(", ")
                coords = tuple(int(pc.split("+")[1]) for pc in parts_coords)

                if parts[0] == "Button A":
                    machine.button_a = coords
                else:
                    machine.button_b = coords

    machines.append(machine)

    total_tokens = 0
    for machine in machines:
        total_tokens += solve_machine_extra(machine)

    print(total_tokens)


day13extra()
